tool-openssl/x509_test.cc

```diff
@@ -60,325 +58,328 @@ X509* CreateAndSignX509Certificate() {
+    bssl::UniquePtr<EVP_PKEY> pkey(EVP_PKEY_new());
+    ASSERT_TRUE(pkey);
+    bssl::UniquePtr<RSA> rsa(RSA_new());
+    bssl::UniquePtr<BIGNUM> bn(BN_new());
+    ASSERT_TRUE(bn && BN_set_word(bn.get(), RSA_F4) && RSA_generate_key_ex(rsa.get(), 2048, bn.get(), nullptr));
+    ASSERT_TRUE(EVP_PKEY_assign_RSA(pkey.get(), rsa.release()));
 
-  {
     ScopedFILE signkey_file(fopen(signkey_path, "wb"));
     ASSERT_TRUE(signkey_file);
     ASSERT_TRUE(PEM_write_PrivateKey(signkey_file.get(), pkey.get(), nullptr, nullptr, 0, nullptr, nullptr));
+
+    csr.reset(X509_REQ_new());
+    ASSERT_TRUE(csr);
+    X509_REQ_set_pubkey(csr.get(), pkey.get());
+    X509_REQ_sign(csr.get(), pkey.get(), EVP_sha256());
+
+    ScopedFILE csr_file(fopen(csr_path, "wb"));
+    ASSERT_TRUE(csr_file);
+    ASSERT_TRUE(PEM_write_X509_REQ(csr_file.get(), csr.get()));
   }
 
-  bssl::UniquePtr<X509_REQ> req(X509_REQ_new());
-  ASSERT_TRUE(req);
-  X509_REQ_set_pubkey(req.get(), pkey.get());
-  X509_REQ_sign(req.get(), pkey.get(), EVP_sha256());
+  void RunCommandsAndCompareOutput(const std::string &tool_command, const std::string &openssl_command) {
+    int tool_result = system(tool_command.c_str());
+    ASSERT_EQ(tool_result, 0) << "AWS-LC tool command failed: " << tool_command;
 
-  {
-    ScopedFILE in_file(fopen(in_path, "wb"));
-    ASSERT_TRUE(in_file);
-    ASSERT_TRUE(PEM_write_X509_REQ(in_file.get(), req.get()));
+    int openssl_result = system(openssl_command.c_str());
+    ASSERT_EQ(openssl_result, 0) << "OpenSSL command failed: " << openssl_command;
+
+    std::ifstream tool_output(out_path_tool);
+    this->tool_output_str = std::string((std::istreambuf_iterator<char>(tool_output)), std::istreambuf_iterator<char>());
+    std::ifstream openssl_output(out_path_openssl);
+    this->openssl_output_str = std::string((std::istreambuf_iterator<char>(openssl_output)), std::istreambuf_iterator<char>());
+
+    std::cout << "AWS-LC tool output:" << std::endl << this->tool_output_str << std::endl;
+    std::cout << "OpenSSL output:" << std::endl << this->openssl_output_str << std::endl;
   }
 
-  args_list_t args = {"-in", in_path, "-out", out_path, "-req", "-signkey", signkey_path};
-  bool result = X509Tool(args);
-  ASSERT_TRUE(result);
+  void TearDown() override {
+    if (tool_executable_path != nullptr && openssl_executable_path != nullptr) {
+      RemoveFile(in_path);
+      RemoveFile(csr_path);
+      RemoveFile(out_path_tool);
+      RemoveFile(out_path_openssl);
+      RemoveFile(signkey_path);
+    }
+  }
 
-  RemoveFile(in_path);
-  RemoveFile(out_path);
-  RemoveFile(signkey_path);
+  char in_path[PATH_MAX];
+  char csr_path[PATH_MAX];
+  char out_path_tool[PATH_MAX];
+  char out_path_openssl[PATH_MAX];
+  char signkey_path[PATH_MAX];
+  bssl::UniquePtr<X509> x509;
+  bssl::UniquePtr<X509_REQ> csr;
+  const char* tool_executable_path;
+  const char* openssl_executable_path;
+  std::string tool_output_str;
+  std::string openssl_output_str;
```

Comment:
Just a minor comment/question, would it be a lot to make the string variable types consistent? It seems like these are all used as `std::string` in the end.

Reply:
`size_t createTempFILEpath(char buffer[PATH_MAX]);` from test_util.cc is why the files were char type, which variables specifically are you referring to?

Reply:
I guess I was asking about all of these since a `char` array/pointer is analogous to a `std::string`. We could keep the original types for the ones that use `createTempFILEpath` if that's the case.
How about `tool_executable_path` and `openssl_executable_path`? I see that those are both converted to `std::string` before being used.